Paper writing (#11)
esa/polyhedral-gravity-model: paper/paper.md

# Polyhedral Model

On a mathematical level, the implemented model follows the approach by Petrović [@petrovic1996determination] as refined by Tsoulis and Petrović [@tsoulis2001singularities]. A comprehensive description of the mathematical foundations of the model is given in the associated student report [@jonas_report]. 

Implementation-wise it makes use of the inherent parallelization opportunity of the approach as it iterates over the mesh. This parallelization is achieved via *thrust* which allows utilizing *OpenMP* and *Intel TBB*. On a finer scale, individual costly operations were investigated and, e.g., the \texttt{arctan} operations were vectorized using *xsimd*. On an application side, the user may use the implemented caching mechanism to avoid recomputation of mesh properties, such as the face normals.

Extensive tests using GoogleTest are used via GitHub Actions to ensure the (continued) correctness of the implementation.

# Installation \& Contribution

> The library is available on GitHub \footnote{\url{https://github.com/esa/polyhedral-gravity-model}} and can be installed with *pip* \footnote{\url{https://github.com/esa/polyhedral-gravity-model#pip}} or from *conda* \footnote{\url{https://anaconda.org/conda-forge/polyhedral-gravity-model}}. Build instructions using *CMake* are provided in the repository. The library is licensed under a GPL license.

Author: > can be installed with *pip* \footnote{\url{https://github.com/esa/polyhedral-gravity-model#pip}}

@schuhmaj Did we ever consider properly putting it on pip? Would be nicest if people can just `pip install polyhedral-gravity-model`

---
title: 'Efficient Polyhedral Gravity Modeling in Modern C++ and Python'
tags:
  - C++
  - Python
  - astronomy
  - dynamics
  - asteroids
  - gravity
  - numerical methods
  - polyhedral model
authors:
  - name: Jonas Schuhmacher
    orcid: 0009-0005-9693-4530
    affiliation: 1
  - name: Emmanuel Blazquez
    orcid: 0000-0001-9697-582X
    affiliation: 2
  - name: Fabio Gratl
    orcid: 0000-0001-5195-7919
    affiliation: 1
  - name: Dario Izzo
    orcid: 0000-0002-9846-8423
    affiliation: 2
  - name: Pablo Gómez
    orcid: 0000-0002-5631-8240
    affiliation: 2
affiliations:

>  - name: Technische Universität München, Arcisstraße 21, 80333 München, Germany 

Reviewer: ```suggestion
 - name: Chair for Scientific Computing, Technische Universität München, Arcisstraße 21, 80333 München, Germany 
```

   index: 1
 - name: Advanced Concepts Team, European Space Agency, European Space Research and Technology Centre (ESTEC), Keplerlaan 1, 2201 AZ Noordwijk, The Netherlands
   index: 2
date: 07 July 2023
bibliography: paper.bib

---

# Summary

> Polyhedral gravity models are ubiquitous for modeling the gravitational field of irregular bodies, such as asteroids and comets. We present an open-source C++ library for the efficient, parallelized computation of a polyhedral gravity model. We also provide a Python interface to the library using *pybind11*. The library is particularly focused on delivering high performance and scalability which we achieve through vectorization and parallelization with *xsimd* and *thrust*, respectively. The library supports many common formats, such as *.stl*, *.off*, *.ply*, *.mesh* and *tetgen*'s *.node* and *.face*.

Reviewer: ```suggestion
Polyhedral gravity models are ubiquitous for modelling the gravitational field of irregular bodies, such as asteroids and comets. We present an open-source C++ library for the efficient, parallelized computation of a polyhedral gravity model. We also provide a Python interface to the library using *pybind11*. The library is particularly focused on delivering high performance and scalability, which we achieve through vectorization and parallelization with *xsimd* and *thrust*, respectively. The library supports many common formats, such as *.stl*, *.off*, *.ply*, *.mesh* and *tetgen*'s *.node* and *.face*.
```
US or British English?

Author: Yes

> Polyhedral gravity models are ubiquitous for modelling the gravitational field of irregular bodies, such as asteroids and comets. We present an open-source C++ library for the efficient, parallelized computation of a polyhedral gravity model. We also provide a Python interface to the library using *pybind11*. The library is particularly focused on delivering high performance and scalability, which we achieve through vectorization and parallelization with *xsimd* and *thrust*, respectively. The library supports many common formats, such as *.stl*, *.off*, *.ply*, *.mesh* and *tetgen*'s *.node* and *.face*.

Reviewer 2: Good point 👍

# Statement of Need

> The complex gravitational fields of irregular bodies, such as asteroids and comets, are often modeled using polyhedral gravity models as they provide an analytic solution for the computation of the gravitational potential, acceleration (and second derivative) given a mesh of the body [@tsoulis2012analytical;@tsoulis2021computational]. The computation of the gravitational potential and acceleration is a computationally expensive task, especially for large meshes, which can however benefit from parallelization either over computed targets points for which we seek potential and acceleration or over the mesh. Thus, a high-performance implementation of a polyhedral gravity model is desirable. 

Reviewer: ```suggestion
The complex gravitational fields of irregular bodies, such as asteroids and comets, are often modeled using polyhedral gravity models as they provide an analytic solution for the computation of the gravitational potential, acceleration (and second derivative) given a mesh of the body [@tsoulis2012analytical;@tsoulis2021computational]. The computation of the gravitational potential and acceleration is a computationally expensive task, especially for large meshes, which can however benefit from parallelization either over computed target points for which we seek potential and acceleration or over the mesh. Thus, a high-performance implementation of a polyhedral gravity model is desirable. 
```

> The complex gravitational fields of irregular bodies, such as asteroids and comets, are often modeled using polyhedral gravity models as they provide an analytic solution for the computation of the gravitational potential, acceleration (and second derivative) given a mesh of the body [@tsoulis2012analytical;@tsoulis2021computational]. The computation of the gravitational potential and acceleration is a computationally expensive task, especially for large meshes, which can however benefit from parallelization either over computed target points for which we seek potential and acceleration or over the mesh. Thus, a high-performance implementation of a polyhedral gravity model is desirable. 

Reviewer 2: Included now:

- Why other models are failing/ unsuitable for the use case (spherical harmonics, mascon models)
- Shifted the focus more on the whole implementation's advantages since we're lacking time measurements for other implementations. Other open source impl. are either a pain to set up. Or papers describing the performance of their employed polyhedral (most of the time, just a sidenote as the model isn't the core of the paper) models lack the required information about the full methodology. I've often seen something like: "Evaluation for one point was expensive and took several seconds," but then they lacked information about the employed strategy (Werner, Tsoulis, etc.), the employed hardware, and the employed mesh.

> While some research code for these models exists, they are not focused on usability and limited to FORTRAN **TODO LINK** and proprietary software like MATLAB **TODO LINK**. There is a lack of well-documented, maintained open-source implementations, particularly in modern programming languages and with a focus on scalability and performance.

Reviewer: ```suggestion
While some research code for these models exists, they are not focused on usability and are limited to FORTRAN **TODO LINK** and proprietary software like MATLAB **TODO LINK**. There is a lack of well-documented, actively maintained open-source implementations, particularly in modern programming languages, and with a focus on scalability and performance.
```

> The presented software has already seen application in several research works. It has been used to optimize trajectories around the highly irregular comet 67P/Churyumov-Gerasimenko [@marak2023trajectory]. Further, it has been used to study the effectiveness of so-called neural density fields [@izzo2022geodesy], where it can serve as a ground truth and to pretrain neural networks [@GeodesyNetsBenchmark]. **TODO_add_more_examples**

Reviewer: ```suggestion
The presented software has already seen application in several research works. It has been used to optimize trajectories around the highly irregular comet 67P/Churyumov-Gerasimenko [@marak2023trajectory]. Further, it has been used to study the effectiveness of so-called neural density fields [@izzo2022geodesy], where it can serve as ground truth and to pre-train neural networks [@GeodesyNetsBenchmark]. **TODO_add_more_examples**
```

Thus, overall this model is highly versatile and can be used in a wide range of applications. We hope it will enable further research in the field, especially related to recent machine learning techniques, which typically rely on Python implementations.

# Polyhedral Model

> On a mathematical level, the implemented model follows the approach by Petrović [@petrovic1996determination] as refined by Tsoulis and Petrović [@tsoulis2001singularities]. A comprehensive description of the mathematical foundations of the model is given in the associated student report [@jonas_report]. 

Reviewer 3: Wouldn't it be better here to cite the ICATT poster / paper, if there's a reference for that?

Reviewer 2: The GNC poster/ paper does (due to size restrictions) only scratches on the underlying math. The details are only in the original paper/ the implementation's technical report.

> Implementation-wise it makes use of the inherent parallelization opportunity of the approach as it iterates over the mesh. This parallelization is achieved via *thrust* which allows utilizing *OpenMP* and *Intel TBB*. On a finer scale, individual costly operations were investigated and, e.g., the \texttt{arctan} operations were vectorized using *xsimd*. On an application side, the user may use the implemented caching mechanism to avoid recomputation of mesh properties, such as the face normals.

Reviewer: ```suggestion
Implementation-wise, it makes use of the inherent parallelization opportunity of the approach as it iterates over the mesh. This parallelization is achieved via *thrust*, which allows utilizing *OpenMP* and *Intel TBB*. On a finer scale, individual, costly operations were investigated and, e.g., the \texttt{arctan} operations were vectorized using *xsimd*. On an application side, the user may use the implemented caching mechanism to avoid the recomputation of mesh properties, such as the face normals.
```

Extensive tests using GoogleTest are used via GitHub Actions to ensure the (continued) correctness of the implementation.

> 

Reviewer 2: Not in that sense. We have figures for
* comparing the different backends (Serial, OpenMP, TBB) on Windows, macOS, Linux
* comparing the impact of the explicit `arctan` vectorization

However, from the theoretical stance, given that each face's computation is independent, the implementation should scale pretty well (of course, given that the number of faces is big enough so that additional cores don't bring more "creation/joining" overhead than we have actually computation work)

# Installation \& Contribution

> The library is available on GitHub \footnote{\url{https://github.com/esa/polyhedral-gravity-model}} and can be installed with *pip* or from *conda* \footnote{\url{https://anaconda.org/conda-forge/polyhedral-gravity-model}}. Build instructions using *CMake* are provided in the repository. The library is licensed under a GPL license.

Reviewer 3: Maybe stupid question but is it available / tested for all relevant operating systems ?

Reviewer 2: Available: Yes: Windows 64, Linux x64 and ARM64 (wheel only for x64 for the recent version 2.1), OSX x64 and ARM64 (wheels only for the x64 for the recent version 2.1) - for each CPython since 3.6 and PyPy since 3.7

Tested: Import Checks are conducted for everything—proper CI Testing only on Linux due to performance reasons. However, I ran the test irregularly in my setup on Windows and OSX.

The project is open to contributions via pull requests with instructions on how to contribute provided in the repository.

# Usage Instructions

> We provide detailed usage instructions in the technical documentation on ReadTheDocs \footnote{\url{https://polyhedral-gravity-model-cpp.readthedocs.io/en/latest/}}. Additionally, a minimal working example is given in the repository readme and more extensive examples as a *Jupyter* notebook \footnote{\url{https://github.com/esa/polyhedral-gravity-model/blob/main/script/polyhedral-gravity.ipynb}}.

Reviewer 3: Is it worth mentioning Rasmus's work here as a companion work with more complexity in application wrt the minimal working example ? It might lead to a review of that repo though so idk.

Reviewer 2: I would argue that the mention in **Statement of Need** is probably sufficient as and would rather focus on the "core" in the chapter **Usage Instructions**.

However, still a good suggestion, I would extend the description of Rasmus's work in **Statement of Need** a bit. The more complex application using pygmo is for sure worth mentioning also regarding the newest addition of `pickling ` for the `GravityEvaluable`

Reviewer 3: Agreed!

![UML Component Diagram of the implementation. External dependencies are depicted in blue. Internal components are colored in grey.\label{fig:implementation}](figures/PolyhedralGravityModel.png)

Extensive tests using GoogleTest for the C++ side and pytest for the Python interface are employed via GitHub Actions to ensure the (continued) correctness of the implementation.

\autoref{fig:implementation} summarizes the modular implementation and its dependencies in a UML component diagram.

# Installation \& Contribution

The library is available on GitHub\footnote{\url{https://github.com/esa/polyhedral-gravity-model}} and can be installed with *pip* (PyPi)\footnote{\url{https://pypi.org/project/polyhedral-gravity/}} or from *conda*\footnote{\url{https://anaconda.org/conda-forge/polyhedral-gravity-model}}. Build instructions using *CMake* are provided in the repository. The library is licensed under a GPL license.

The project is open to contributions via pull requests, with instructions on how to contribute provided in the repository.

# Usage Instructions

> We provide detailed usage instructions in the technical documentation on ReadTheDocs\footnote{\url{https://polyhedral-gravity-model-cpp.readthedocs.io/en/latest/}}. Additionally, a minimal working example is given in the repository readme, and more extensive examples, including a walkthrough over the available options as a *Jupyter* notebook\footnote{\url{https://github.com/esa/polyhedral-gravity-model/blob/main/script/polyhedral-gravity.ipynb}}.

Author: @schuhmaj I don't think we added the minimal working example to the readme, did we?

Reviewer 2: Indeed 👍
